src/compiler/checker.ts

```diff
@@ -39034,12 +39034,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return false;
     }
 
-    function mergeInferences(target: InferenceInfo[], source: InferenceInfo[]) {
+    function tryInferencesMerge(target: InferenceInfo[], source: InferenceInfo[]) {
+        let merged = false;
+
+        const areOverlapping = hasOverlappingInferences(target, source);
+
         for (let i = 0; i < target.length; i++) {
-            if (!hasInferenceCandidates(target[i]) && hasInferenceCandidates(source[i])) {
+            if (areOverlapping ? hasInferenceCandidates(source[i]) && hasInferenceCandidates(target[i]) && source[i].priority! < (target[i].priority! & ~InferencePriority.ReturnType) : hasInferenceCandidates(source[i])) {
                 target[i] = source[i];
+                target[i].priority = InferencePriority.None;

```

Comment:
TODO for myself: 

this line might not be needed at all or the assigned priority should be the source priority without `InferencePriority.ReturnType` (the latter sounds like a good idea if those inferences should stay "upgradable"/mergeable). IIUC, no new inferences without `InferencePriority.ReturnType` can be made here anyway - all of those were already collected before the ones from generic signatures start to be included.

Reply:
Hm, yeah, `None` definitely doesn't feel correct, since `None` is actually the highest priority inference possible - and I don't think simply adopting `source`'s inference for `target` when they overlap should guarantee this is the best inference possible across all potential inference sites.

But also - this is mutating the inference object, which you should avoid, since it means if that inference is reused in multiple contexts, you'll be mangling the priority in all of them (...we probably shouldn't be reusing an inference in multiple contexts, but there are probably some edge cases we decided it was OK in that this might break the invariants of). If we can't reuse the `source` inference wholesale (because its priority already doesn't contain `ReturnType`), then we should duplicate the `source` inference object with a new priority.

Reply:
I initially picked `None` here because it's the most conservative thing to do. If I'm reasoning about it correctly. Right now (on `main`) we iterate over those generic signatures (conceptually at least):
- if the inferences are overlapping then we don't try to use any of the new ones
- if the inferences are not overlapping then we use the new ones **but** that means that when we overlap again we won't pick any new ones

So my conservative change adjusts this only slightly - it only allows a single "upgrade" when inferences are overlapping (and in such a case, we only update the overlapping one, this is important - we can't assign inferences that are not overlapping).

I thought that this, both, is just the smallest change that could be done here if this direction is promising at all and that there might be an argument to be made here that in this scenario the inferences made from earlier generic signatures should be preferred (left ones having the higher priority over the right ones). The latter doesn't match how "regular" inferences are gathered but it is how it implicitly works on `main` (when overlap happens the new inferences are ignored after all, so they can't be replaced/updated).

I could certainly experiment with relaxing this behavior and allow those upgrades continuously.

---

I pushed out a change using `cloneInferenceInfo` but I'm not sure if you meant that or `cloneInferenceContext`. And perhaps it should also be applied conditionally? but then that would imply that `cloneInferenceInfo` is the right choice here since `ReturnType` is relevant to the info and not to the whole context. 

> If we can't reuse the source inference wholesale (because its priority already doesn't contain ReturnType)

Does it mean that when the inference info of the source contains `ReturnType` then it's OK to reuse it and avoid cloning?

Reply:
> Since those inferredTypeParameters are simply concatenated there is no good way of even rejecting the unused ones~ at some point.

Yeah, at some point a filtering of that list to remove type parameters that definitely aren't used and logic for figuring out if the type parameter was actually used is probably needed for usability. It'll be a bit of a challenge, though, with all the deferred inferences we can make "maybe it's used later 🤷‍♂️" is probably a somewhat common case.